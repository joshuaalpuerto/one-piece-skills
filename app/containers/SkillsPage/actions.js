import { createActions } from 'redux-actions'

import { GET_SKILLS, CREATE_SKILLS, RESULT_SKILLS } from './constants'

const ReduxActions = createActions({}, GET_SKILLS, CREATE_SKILLS, RESULT_SKILLS)

export const { getSkills, createSkills, resultSkills } = {
  getSkills: ReduxActions['app/skillsPage/getSkills'],
  createSkills: ReduxActions['app/skillsPage/createSkills'],
  resultSkills: ReduxActions['app/skillsPage/resultSkills']
}
