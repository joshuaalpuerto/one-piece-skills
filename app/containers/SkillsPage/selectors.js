/**
 * Homepage selectors
 */

import { createSelector } from 'reselect'
import { initialState } from './reducer'

const selectSkillsDomain = state => state.get('skillsPage', initialState)

const selectSkills = () =>
  createSelector(selectSkillsDomain, subState => subState.get('skills'))

const selectSkillsLoader = () =>
  createSelector(selectSkillsDomain, subState => subState.get('skillsLoader'))

const selectSkillsSuccess = () =>
  createSelector(selectSkillsDomain, subState => subState.get('skillsSuccess'))

const selectSkillsError = () =>
  createSelector(selectSkillsDomain, subState => subState.get('skillsError'))

export {
  selectSkills,
  selectSkillsLoader,
  selectSkillsSuccess,
  selectSkillsError
}
