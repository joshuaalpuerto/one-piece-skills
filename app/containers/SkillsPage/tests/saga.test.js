import { put, select } from 'redux-saga/effects'
import { fromJS } from 'immutable'
import { resultSkills } from '../actions'
import { selectSkills } from '../selectors'

import skillsPageSaga, {
  getSkillsSaga,
  getSkills,
  getResult,
  createSkillsSaga,
  createSkills
} from '../saga'

const skill = {
  name: 'React',
  experience: '<1 year'
}

const args = {
  payload: skill
}

describe('SkillsPage Saga', () => {
  describe('getSkillsSaga', () => {
    beforeEach(() => {
      const generator = getSkillsSaga()
      const latest = generator.next().value
      expect(latest).toMatchSnapshot()
    })

    describe('getSkills [SUCCESS]', () => {
      const generator = getSkills()

      it('should request call to API', () => {
        const call = generator.next().value
        expect(call).toMatchSnapshot()
      })

      it('should set resultRegistration', () => {
        const skills = []
        const callAction = generator.next(skills).value
        expect(callAction).toEqual(getResult(skills))
      })
    })

    describe('getSkills [FAIL]', () => {
      const generator = getSkills(args)

      it('should request call to API', () => {
        const call = generator.next().value
        expect(call).toMatchSnapshot()
      })

      it('should throw error resultSkills', () => {
        const error = new Error()
        const response = {
          response: {
            json() {}
          }
        }
        // eslint-disable-next-line no-unused-vars
        const throwAction = generator.throw(response).value
        const callAction = generator.next(response).value
        expect(callAction).toEqual(put(resultSkills(error)))
      })
    })
  })

  describe('createSkillsSaga', () => {
    beforeEach(() => {
      const generator = createSkillsSaga()
      const latest = generator.next().value
      expect(latest).toMatchSnapshot()
    })

    describe('createSkills [SUCCESS]', () => {
      const generator = createSkills(args)

      it('should request call to API', () => {
        const call = generator.next().value
        expect(call).toMatchSnapshot()
      })

      it('should get skills from selector', () => {
        const selector = generator.next(skill).value

        expect(JSON.stringify(selector)).toEqual(
          JSON.stringify(select(selectSkills()))
        )
      })

      it('should return updated skill list', () => {
        const skills = fromJS([])
        const result = generator.next(skills).value
        expect(result).toEqual(getResult([skill]))
      })
    })

    describe('createSkills [FAIL]', () => {
      const generator = createSkills(args)

      it('should request call to API', () => {
        const call = generator.next().value
        expect(call).toMatchSnapshot()
      })

      it('should throw error resultSkills', () => {
        const error = new Error()
        const response = {
          response: {
            json() {}
          }
        }
        // eslint-disable-next-line no-unused-vars
        const throwAction = generator.throw(response).value
        const callAction = generator.next(response).value
        expect(callAction).toEqual(put(resultSkills(error)))
      })
    })
  })

  describe('defaultGenerators', () => {
    const generators = skillsPageSaga()

    it('should have 2 watchers', () => {
      const watchers = generators.next().value
      expect(watchers.length).toEqual(2)
    })
  })
})
