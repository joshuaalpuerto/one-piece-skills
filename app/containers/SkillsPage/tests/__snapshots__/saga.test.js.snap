// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkillsPage Saga createSkillsSaga createSkills [FAIL] should request call to API 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/CREATE_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga createSkillsSaga createSkills [FAIL] should request call to API 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "undefined/skills",
      Object {
        "body": "{\\"name\\":\\"React\\",\\"experience\\":\\"<1 year\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga createSkillsSaga createSkills [FAIL] should throw error resultSkills 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/CREATE_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga createSkillsSaga createSkills [SUCCESS] should get skills from selector 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/CREATE_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga createSkillsSaga createSkills [SUCCESS] should request call to API 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/CREATE_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga createSkillsSaga createSkills [SUCCESS] should request call to API 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "undefined/skills",
      Object {
        "body": "{\\"name\\":\\"React\\",\\"experience\\":\\"<1 year\\"}",
        "headers": Object {
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga createSkillsSaga createSkills [SUCCESS] should return updated skill list 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/CREATE_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga getSkillsSaga getSkills [FAIL] should request call to API 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/GET_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga getSkillsSaga getSkills [FAIL] should request call to API 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "undefined/skills",
      Object {
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga getSkillsSaga getSkills [FAIL] should throw error resultSkills 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/GET_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga getSkillsSaga getSkills [SUCCESS] should request call to API 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/GET_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga getSkillsSaga getSkills [SUCCESS] should request call to API 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "undefined/skills",
      Object {
        "method": "GET",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`SkillsPage Saga getSkillsSaga getSkills [SUCCESS] should set resultRegistration 1`] = `
Object {
  "@@redux-saga/IO": true,
  "FORK": Object {
    "args": Array [
      "app/SkillsPage/GET_SKILLS",
      [Function],
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
